casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -174,69 +249,130 @@ class RuntimeManagerImpl[F[_]: Concurrent] private[rholang] (
+      computeEffect(runtime)(ConstructDeploy.sourceDeployNow(s"""for(_<-@"$name"){ Nil }"""))
 
   def getContinuation(
-      hash: ByteString,
-      channels: immutable.Seq[Par]
-  ): F[Seq[(Seq[BindPattern], Par)]] =
-    withResetRuntime(
-      hash,
-      runtime => {
-        runtime.space
-          .getWaitingContinuations(channels)
-          .map(
-            results =>
-              for {
-                result <- results.filter(_.continuation.taggedCont.isParBody)
-              } yield (result.patterns, result.continuation.taggedCont.parBody.get.body)
-          )
-      }
+      start: StateHash
+  )(channels: Seq[Par]): F[Seq[(Seq[BindPattern], Par)]] =
+    withResetRuntime(start)(
+      _.space
+        .getWaitingContinuations(channels)
+        .map(
+          _.filter(_.continuation.taggedCont.isParBody)
+            .map(result => (result.patterns, result.continuation.taggedCont.parBody.get.body))
+        )
     )
 
   private def newEval(
-      terms: Seq[DeployData],
-      runtime: Runtime[F],
-      initHash: StateHash
-  ): F[(StateHash, Seq[InternalProcessedDeploy])] = {
-
-    def doEval(
-        terms: Seq[DeployData],
-        hash: Blake2b256Hash,
-        acc: Seq[InternalProcessedDeploy]
-    ): F[(StateHash, Seq[InternalProcessedDeploy])] =
-      Concurrent[F].defer {
-        terms match {
-          case deploy +: rem =>
-            for {
-              _ <- runtime.space.reset(hash)
-              (codeHash, phloPrice, userId, timestamp) = ProtoUtil.getRholangDeployParams(
-                deploy
-              )
-              _                                      <- runtime.shortLeashParams.setParams(codeHash, phloPrice, userId, timestamp)
-              injResult                              <- doInj(deploy, runtime.reducer, runtime.errorLog)(runtime.cost)
-              EvaluateResult(evaluationCost, errors) = injResult
-              newCheckpoint                          <- runtime.space.createCheckpoint()
-              deployResult = InternalProcessedDeploy(
-                deploy,
-                Cost.toProto(evaluationCost),
-                newCheckpoint.log,
-                DeployStatus.fromErrors(errors)
-              )
-              cont <- if (errors.isEmpty)
-                       doEval(rem, newCheckpoint.root, acc :+ deployResult)
-                     else doEval(rem, hash, acc :+ deployResult)
-            } yield cont
-
-          case _ => (ByteString.copyFrom(hash.bytes.toArray), acc).pure[F]
+      runtime: Runtime[F]
+  )(deploys: Seq[DeployData], withPayment: Boolean): F[(StateHash, Seq[InternalProcessedDeploy])] =
+    Concurrent[F]
+      .defer {
+        runtime.space.createCheckpoint() >>= { checkpoint0 =>
+          deploys.toList.foldM((checkpoint0.root, Seq.empty[InternalProcessedDeploy])) {
+            case ((prev, processedDeploys), deploy) =>
+              import deploy._
+              if (withPayment) {
```

Comment:
I'm fine with that. The reason it's not used in production code is because `replayComputeState` does not yet replay the payment. Thus, the replay state would never match the computed state.